### PR TITLE
execNewBlock: remove the parent header as parameter and sync from the db instead

### PR DESCRIPTION
**Pull request by ak3n**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205566925523372

### CODE REVIEW

File: test/Chainweb/Test/Pact/PactMultiChainTest.hs [hunk 1]
```diff
@@ -1404,6 +1418,12 @@ getPWO chid = do
   pwo <- liftIO $ casLookupM pdb (_blockPayloadHash h)
   return (pwo,h)
 
+getPWOByHeader :: BlockHeader -> PactTestM PayloadWithOutputs
+getPWOByHeader h = do
+  (TestBlockDb _ pdb _) <- view menvBdb
```
Comment (edmundnoble) [inline test/Chainweb/Test/Pact/PactMultiChainTest.hs:1423]: No need for these parens

**Review by edmundnoble — COMMENTED** [review]
This can't be merged until the release. All of my comments are pedantic stuff which barely matters, I'm extremely impressed that we can get away with this, it looks exactly like I'd expect. We need to test this on a real mining node as well before mergew.

File: test/Chainweb/Test/Pact/PactMultiChainTest.hs [hunk 2]
```diff
@@ -1404,6 +1418,12 @@ getPWO chid = do
...
+  pwo <- liftIO $ casLookupM pdb (_blockPayloadHash h)
+  return pwo
```
[the remark this answers is not in the document]
Comment (edmundnoble) [reply test/Chainweb/Test/Pact/PactMultiChainTest.hs:1425]: Hell maybe this should be a shared utility.

**Comment by edmundnoble** [comment]
We actually don't really need to test this on a real mining node, fake mining with a simulated time distribution also works, as long as it's in a network with multiple mining nodes.

**Comment by ak3n** [comment]
@edmundnoble will wait until the master is fixed so we can merge with green CI this PR? Or can merge already? it was green before

**Comment by edmundnoble** [comment]
@ak3n I want this PR benchmarked to see how many blocks we produce with a simulated miner first. I can help with that some time when you're free.

**Comment by edmundnoble** [comment]
Actually we should probably use a load test for this benchmark.

File: bench/Chainweb/Pact/Backend/ForkingBench.hs [hunk 3]
```diff
@@ -129,24 +129,40 @@ bench rdb = C.bgroup "PactService"
     forkingBench = withResources rdb 10 Quiet
         $ \mainLineBlocks pdb bhdb nonceCounter pactQueue _ ->
             C.bench "forkingBench"  $ C.whnfIO $ do
-              let (T3 _ join1 _) = mainLineBlocks !! 5
+              let (T3 _ join1 pwo1) = mainLineBlocks !! 5
+
+              -- we have to execValidateBlock on `join1` block height to update the parent header
+              void $ validateBlock join1 (payloadWithOutputsToPayloadData pwo1) pactQueue
```
Comment (0xd34df00d) [inline bench/Chainweb/Pact/Backend/ForkingBench.hs:135]: You might create a helper function (for the tests only!) like
```haskell
updateParentHeader join pwo queue = void $ validateBlock join (payloadWithOutputsToPayloadData pwo) queue
```
avoiding the comments in the tests and making the tests a bit more expressive.

File: src/Chainweb/Pact/PactService.hs [hunk 4]
```diff
@@ -292,9 +292,9 @@ serviceRequests memPoolAccess reqQ = do
                 go
             NewBlockMsg NewBlockReq {..} -> do
                 trace logFn "Chainweb.Pact.PactService.execNewBlock"
-                    (_parentHeader _newBlockHeader) 1 $
-                    tryOne "execNewBlock" _newResultVar $
-                        execNewBlock memPoolAccess _newBlockHeader _newMiner
+                    () 1 $
+                    tryOne "execNewBlock" _newResultVar $ do
```
Comment (0xd34df00d) [inline src/Chainweb/Pact/PactService.hs:296]: Tiny nit: this `do` isn't necessary.

File: src/Chainweb/Pact/Service/Types.hs [hunk 5]
```diff
@@ -292,11 +292,10 @@ data RequestMsg = NewBlockMsg !NewBlockReq
 type PactExMVar t = MVar (Either PactException t)
 
 data NewBlockReq = NewBlockReq
-    { _newBlockHeader :: !ParentHeader
-    , _newMiner :: !Miner
+    { _newMiner :: !Miner
     , _newResultVar :: !(PactExMVar PayloadWithOutputs)
     }
-instance Show NewBlockReq where show NewBlockReq{..} = show (_newBlockHeader, _newMiner)
+instance Show NewBlockReq where show NewBlockReq{..} = show (_newMiner)
```
Comment (0xd34df00d) [inline src/Chainweb/Pact/Service/Types.hs:298]: The parens are probably no longer needed.

**Review by 0xd34df00d — APPROVED** [review]
The logic looks good as per discussion, so I only have a few minor nits.

**Comment by ak3n** [comment]
Merged in #1803

**Pull request closed without merging**